### ISSUE TITLE
torch.distributed.pipelining _split_block_mask hides batch offset from non-strict tracing

### ISSUE THREAD

**Issue by sanketpurandare** [pytorch/pytorch#188727]
## Issue

`torch.distributed.pipelining.microbatch._split_block_mask` creates split `BlockMask` objects whose tensor metadata is correctly split, but whose generated `mask_mod` captures the per-split batch offset as a Python `int` closure value.

That is fine for eager execution, but it is not a good tracing contract. `BlockMask._flatten()` is designed to expose closure tensor leaves to tracing. Since the split offset is a Python `int`, non-strict tracing/export specializes it instead of representing it as replay-varying input state.

The result is that a graph traced/exported with split 0 can replay split 1 with offset 0.

## Repro

This is self-contained and uses only PyTorch APIs.

```python
import torch
from torch import nn
from torch.distributed.pipelining.microbatch import _split_block_mask
from torch.nn.attention.flex_attention import (
    BlockMask,
    _MaskModWrapper,
    create_block_mask,
)
from torch.utils import _pytree as pytree

if BlockMask not in pytree.SUPPORTED_NODES:
    pytree.register_pytree_node(
        BlockMask,
        BlockMask._flatten,
        BlockMask._unflatten,
        flatten_with_keys_fn=BlockMask._flatten_with_keys,
        serialized_type_name="torch.nn.attention.flex_attention.BlockMask",
    )
if _MaskModWrapper not in pytree.SUPPORTED_NODES:
    pytree.register_constant(_MaskModWrapper)


def mask_mod(b, h, q_idx, kv_idx):
    return (b == 1) & (q_idx >= kv_idx)


base = create_block_mask(
    mask_mod,
    B=2,
    H=None,
    Q_LEN=128,
    KV_LEN=128,
    device="cpu",
)
mask0, mask1 = _split_block_mask(base, 2)


class M(nn.Module):
    def forward(self, mask):
        b = torch.tensor(0)
        h = torch.tensor(0)
        q_idx = torch.tensor(1)
        kv_idx = torch.tensor(0)
        return mask.mask_mod(b, h, q_idx, kv_idx)


m = M()
print("eager", m(mask0).item(), m(mask1).item())

ep = torch.export.export(m, (mask0,), strict=False)
print(ep.graph_module.code)

# Call the exported flat graph directly to demonstrate the baked offset.
# The public ep.module() wrapper also rejects mask1 because the BlockMask input
# tree spec differs after tracing, which is another symptom of the same hidden
# Python closure state.
consts = [
    ep.constants["lifted_tensor_0"],
    ep.constants["lifted_tensor_1"],
    ep.constants["lifted_tensor_2"],
]


def split_mask_tensor_leaves(mask):
    return [leaf for leaf in pytree.tree_leaves(mask) if isinstance(leaf, torch.Tensor)][:8]


print(
    "flat graph replay",
    ep.graph_module(*consts, *split_mask_tensor_leaves(mask0))[0].item(),
    ep.graph_module(*consts, *split_mask_tensor_leaves(mask1))[0].item(),
)
```

## Actual result

The eager values are correct:

```text
eager False True
```

But the exported graph specializes the split offset as `0`:

```python
full_like = torch.ops.aten.full_like.default(detach_, 0, pin_memory = False)
add = torch.ops.aten.add.Tensor(detach_, full_like)
```

and replaying the flat graph with split 1 gives the wrong value:

```text
flat graph replay False False
```

## Expected result

Replay-varying state in a split `BlockMask` should be visible to non-strict tracing/export through `BlockMask._flatten()`.

For this repro, split 1 should preserve the batch offset and replay as:

```text
flat graph replay False True
```

## Suggested fix

Change `_split_block_mask` so that the generated `mask_mod` captures the split batch offset as a scalar tensor closure leaf instead of a Python `int`.

Sketch:

```python
def create_mask_mod(idx):
    batch_offset = torch.tensor(
        idx,
        device=block_mask.kv_num_blocks.device,
        dtype=torch.int64,
    )

    def batch_offset_mask_mod(b, h, q_idx, kv_idx):
        offset = batch_offset.to(device=b.device, dtype=b.dtype)
        return block_mask.mask_mod(b + offset, h, q_idx, kv_idx)

    return batch_offset_mask_mod
```

Important detail: the scalar tensor must be a closure leaf, not a default argument, so that `BlockMask._flatten()` can expose it.

This should preserve eager semantics while making the split masks replayable under non-strict tracing/export.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98 @Chillee @drisspg @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv

**Comment by drisspg** [comment]
tbh I dont really care about eager semantics just dyanmo semantics

**Comment by sanketpurandare** [comment]
> tbh I dont really care about eager semantics just dyanmo semantics

@drisspg 
1. This is a torch.distributed.pipelining bug not a flex attention bug.
2. PP splits the local batch into microbatches and that where we encountered this. 
3. Dynamo might be the end goal for flex attention but not for other work streams like GraphPP and Eager Torchtitan PP.
4. We are changing PP to no longer handle microbatch splitting and let the user supply microbatches here: https://github.com/pytorch/pytorch/pull/188500
5. Torchtitan will now produce microbatches right at source i.e. dataloader: https://github.com/pytorch/torchtitan/pull/3856
6. So no changes are required from your end and I had already assigned myself to this issue.
7. If you do not want to support flex_attention outside dynamo, then we should add a hard error in flex_attention not a warning so that people not using full torch.compile will stop using it.

**Comment by drisspg** [comment]
7. I think the error is pretty straightforward; you aint getting flash like impl if you are not going to inductor. If you think there a valid use case for using this impl for eagerPP then ill lean on your judgement here. 

3. end goal is a too strong of a phrase more so we know how to guarantee the invariants with dyanmo, it is harder to do so in eager and areas. All i am saying is that
>This should preserve eager semantics while making the split masks replayable under non-strict tracing/export. 
is not fully correct and the case we care about is producing graphs we know are correct and know when they need to be invalidated which dynamo does and eager does not. 

6. Sounds good  ill let you handle it. I was trying to figure out from your snippet if things like:

```

if BlockMask not in pytree.SUPPORTED_NODES:
    pytree.register_pytree_node(
        BlockMask,
        BlockMask._flatten,
        BlockMask._unflatten,
        flatten_with_keys_fn=BlockMask._flatten_with_keys,
        serialized_type_name="torch.nn.attention.flex_attention.BlockMask",
    )
if _MaskModWrapper not in pytree.SUPPORTED_NODES:
    pytree.register_constant(_MaskModWrapper)


def mask_mod(b, h, q_idx, kv_idx):
    return (b == 1) & (q_idx >= kv_idx)
```

is user code or handled by the framework in which case wanted to understand what things need to be rewritten and by whom

**Comment by sanketpurandare** [comment]
So the _split_block_mask was a utility introduced a while back in PP where it relied on framework to do the splitting for the user. But as more variants of attention like Varlen or let's in future some other thing is introduced it will be hard for the framework to know how to split it. GraphTrainer traces block mask but marks it as a terminal pytree node. On RL side people want to split their microbatches themselves and also use varying seq lengths.

So the goal of this issue was to highlight the fact that framework doing the splitting for arbitrary user args/kwargs is not the right thing to do. Secondly a simple fix that makes the batch offset as a scalar tensor already works and replays the block mask correctly. 

Check how GraphTrainer is able to do all of this with bitwise numerics.

**Comment by sanketpurandare** [comment]
TLDR; Dynamo is not that special, just make everything a tensor, use python primitives only if they are identical across different invocations. JAX uses non-strict model and hence it forced users to change the way they program.
